File: Problemset/house-robber-ii/house-robber-ii.py

# @Title: 打家劫舍 II (House Robber II)
# @Author: cocofe
# @Date: 2021-04-15 23:55:17
# @Runtime: 40 ms
# @Memory: 14.7 MB

class Solution:
    def rob(self, nums: List[int]) -> int:
        def helper(start, end):
            pprev, prev = 0, 0
            for idx in range(start, end+1):
                num = nums[idx]
                curr = max(prev, pprev+num)
                pprev, prev = prev, curr
            return prev
        if len(nums) <= 2:
            return max(nums)
        return max(helper(0, len(nums)-2), helper(1, len(nums)-1))
            

File: Problemset/house-robber-ii/house-robber-ii.java

// @Title: 打家劫舍 II (House Robber II)
// @Author: cocofe
// @Date: 2021-07-24 18:52:42
// @Runtime: 0 ms
// @Memory: 36 MB

class Solution {
    public int rob(int[] nums) {
        return Math.max(helper(nums, 0, nums.length-2), helper(nums, 1, nums.length-1));
    }
    private int helper(int[] nums, int i, int j){
        if (i >= nums.length) {
            return 0;
        }
        int a=0, b = 0, c = nums[i];
        for (int idx=i+1; idx <= j; idx ++) {
            int tmp1 = a, tmp2 = b;
            a = b;
            b = c;
            c = Math.max(tmp1, tmp2) + nums[idx];
            // System.out.printf("idx:%s, a:%s, b:%s, c:%s \n", idx, a, b, c);
        }
        // System.out.printf("i:%s j:%s -> %s\n", i, j, Math.max(b, c));
        return Math.max(b, c);
    }
}
